### PR TITLE
SQL: Improve painless script generated from `IN`

### PR DESCRIPTION
**Pull request by matriv**
Replace standard `||` and `==` painless operators with
`or` and `eq` null safe alternatives from `InternalSqlScriptUtils`.

Follow up to #34750

### CODE REVIEW

**Comment by jdconrad** [comment]
Just a note that '==' in Painless may already cover this use case.  Internally, during a comparison, if either the lhs or rhs is a non-primitive type, `Objects.equals` is called.

**Comment by matriv** [comment]
@jdconrad thx for info! With @costin's guidance we follow now a different approach with implementing our `in` script function so that the logic is still in once place in Java code.

**Comment by matriv** [comment]
retest this please

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/whitelist/InternalSqlScriptUtils.java [hunk 1]
```diff
@@ -113,6 +115,10 @@ public static Boolean notNull(Object expression) {
         return IsNotNullProcessor.apply(expression);
     }
 
+    public static Boolean in(Object value, List<Object> values) {
+        return In.doFold(value, values);
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/whitelist/InternalSqlScriptUtils.java]: The naming convention is to call the method on the processor (typically `InProcessor.apply`)

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java [hunk 2]
```diff
@@ -84,17 +80,21 @@ public boolean foldable() {
...
 
-        Object foldedLeftValue = value.fold();
+        return doFold(value.fold(), Foldables.valuesOf(list, value.dataType()));
+    }
+
+    public static Boolean doFold(Object value, List<Object> values) {
         Boolean result = false;
-        for (Expression rightValue : list) {
-            Boolean compResult = Comparisons.eq(foldedLeftValue, rightValue.fold());
+        for (Object v : values) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java]: Move this method to the processor please.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java [hunk 3]
```diff
@@ -84,17 +80,21 @@ public boolean foldable() {
...
+            Boolean compResult = Comparisons.eq(value, v);
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java]: Let's be explicit about boxing. `Boolean result = Boolean.FALSE`. `if (compResult == Boolean.TRUE) or `compResult.boolean()` ...

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java [hunk 4]
```diff
@@ -122,34 +122,18 @@ public Attribute toAttribute() {
...
-        }
-
-        ParamsBuilder paramsBuilder = paramsBuilder().script(leftScript.params());
-        for (Params p : rightParams) {
-            paramsBuilder = paramsBuilder.script(p);
-        }
-
-        return new ScriptTemplate(format(Locale.ROOT, "%s", sj.toString()), paramsBuilder.build(), dataType());
+        // remove duplicates
+        List<Object> values = new ArrayList<>(
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java]: `new LinkedHashSet(Foldables.valuesOf(list, dataType())`
This has the side effect of triggering a conversion if needed (which should expose potential bugs).

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 5]
```diff
@@ -208,12 +209,11 @@ public void testTranslateInExpression_WhereClause_Painless() {
...
-        assertEquals("InternalSqlScriptUtils.nullSafeFilter(" +
-            "InternalSqlScriptUtils.power(InternalSqlScriptUtils.docValue(doc,params.v0),params.v1)==10 || " +
-            "InternalSqlScriptUtils.power(InternalSqlScriptUtils.docValue(doc,params.v0),params.v1)==20)",
-            sq.script().toString());
-        assertEquals("[{v=int}, {v=2}]", sq.script().params().toString());
+        ScriptQuery sc = (ScriptQuery) translation.query;
+        assertEquals("InternalSqlScriptUtils.nullSafeFilter(InternalSqlScriptUtils.in(" +
+                "InternalSqlScriptUtils.power(InternalSqlScriptUtils.docValue(doc,params.v0),params.v1), params.v2))",
+            sc.script().toString());
+        assertEquals("[{v=int}, {v=2}, {v=[10, null, 20]}]", sc.script().params().toString());
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: 👍

**Review by costin — APPROVED** [review]
Left some minor comments. Looks good!

**Comment by matriv** [comment]
@costin Thanks a lot for the review and all your help on this PR!

**Review by astefan — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
Backported to `6.x` with dab666501e6f6eca4852ba561b9ff65b17fa55fc

**Comment by matriv** [comment]
Backported to `6.5` with c94aeef8f796e9ff16c97797520a9ab6d41aca60

**Pull request merged**